text/0120-js-sdk-scope.md

```diff
@@ -0,0 +1,187 @@
+
+Especially when working with OpenTelemetry, the way how we currently always assign data to the current scope can break down in unexpected ways.
+
+Because OpenTelemetry has to fork a context each time _anything_ is changed on it (copy-on-write), you end up with a lot of deeply nested scopes.
+
+This leads to many scenarios where data is added to the "wrong" scope, from a users perspective:
+
+```js
+app.get('/my-route', function() {
+  db.query(myQuery, function() {
+    console.log('query done!');
+  });
+
+  throw new Error(); // <-- this error would not have the "query done!" breadcrumb!
+});
+```
+
+Since in the example above `db.query()` internally adds spans for the DB operations, which triggeres a new context in OpenTelemetry, nothing that happens inside of this callback would be added to any events outside of the callback. 
+
+Making matters worse, the exact behavior there relies on how auto-instrumentation works under the hood. Depending on what is forked and where, breadcrumbs etc. _may_ be available outside, or not. This is unpredictable and hard to explain.
+
+In order to a) Simplify the concepts our users have to learn about and b) Make interoperability with OpenTelemetry context forking better, this RFC sets out to rework how our scopes work. This is focusing on JavaScript for now, but may be applicable to other languages as well.
+
+# Background
+
+TODO
+
+# Proposed Solution
+
+We introduce a new global (or client) scope, as well as a request scope.
+
+There will be new APIs for this:
+
+* `getGlobalScope()`
+* `getRequestScope()`
+
+Plus an escape hatch for power users to make queues etc. easier to work with:
+
+* `scope.captureBreadcrumbsHere()`
+
+## Breadcrumbs
+
+Breadcrumbs are a key case we want to solve with this RFC. This RFC proposes the following logic for adding/getting breadcrumbs:
+
+* Each scope has a list of breadcrumbs
+* When a breadcrumb is added, it is added based on the following logic:
+  * Is there a "Breadcrumbs Scope" defined? If so, add it there.
+  * Else, is there a "Request Scope" defined? if so, add it there.
+  * Else, add it to the "Global Scope"
+
+This means that e.g. in Browser JS SDK, all breadcrumbs will be global, while in request-based backend SDKs, they will be per-request.
+
+In cases where you want to isolate breadcrumbs manually (e.g. a queue job), you can manually define a breadcrumbs scope.
+
+## Global Scope
+
+First, there is a _Global Scope_. This scope is tied to the client, and anything added to the global scope will always be added to _all_ events.
+Note that the global scope _is not_ the initial scope, but a separate, "empty" scope. This way, you cannot accidentally pollute it. 
+You can access it via `getGlobalScope()`:
+
+```js
+Sentry.init();
+
+const scope = Sentry.getScope(); // <-- the initial scope, that other scopes will inherit from
+const globalScope = Sentry.getGlobalScope(); // <-- the global scope that will be applied to all events
+
+scope !== globalScope
```

Comment:
Does `Sentry.init` create both the `scope` and `globalScope` and set the same client on both of them or only the `globalScope` is tied to the created client and `scope` has `NoopClient`?

Do we always have these two scopes? What would be the difference for envs like RN, where we have currently only one Hub and Scope?

Reply:
This is a bit of an implementation detail, I think I would implement it in a way that the global scope is created when it is first fetched. E.g. something like this:

```js
function getGlobalScope() {
  const client = getClient();
  const globalScope = getGlobalScopeForClient(client); // this could either live on the client, or e.g. in a weakmap
  if (globalScope) {
    return globalScope;
  }
  return createGlobalScopeForClient(client);
}
```

Does that make sense to you? 

I think it may also be reasonable to _not_ have this in RN if we only have a single hub/scope. (Just to clarify, does that mean in RN we do not have `withScope()` or `Scope.clone()`? )

Reply:
RN has `withScope()` or `Scope.clone()` but it's basically not used. One exception is cloning the scope for specific event -> `captureException(error, withScope)`.

Reply:
From https://github.com/getsentry/rfcs/pull/122 I understand that scopes hold reference to clients, 1:1.

Will that be the same with the global scope or it will be available to all created clients?